### PR TITLE
Implement `Store::delete()`

### PR DESCRIPTION
**Pull request by neithernut**
Targets #106.

### CODE REVIEW

File: libimagstore/src/store.rs [hunk 1]
```diff
@@ -147,7 +147,14 @@ impl Store {
 
     /// Delete an entry
     pub fn delete(&self, id: StoreId) -> Result<()> {
-        unimplemented!();
+        let mut entries = self.entries.write();
+
+        // if the entry is currently modified by the user, we cannot drop it
+        if entries.unwrap().get().and_then( |entry| { entry.isBorrowed() } ) {
+            return Result<StoreError::new(IdLocked)>;
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: [Result](https://doc.rust-lang.org/stable/std/result/enum.Result.html) is an enum that has two variants, `Ok` or `Err`. The constructor of each variant is the same as their name. In this case `Result::Err`, however for ease of use Rust already `use`s the `Err` and `Ok` variant for you, thus you just have to write `Err(StoreError::new(..))`.

Comment (neithernut) [inline libimagstore/src/store.rs]: Yep, I already fixed the error locally. Thanks, anyway.

**Comment by neithernut** [comment]
Mh... I should definitely take some time and get me all those crates so I can build locally.

**Comment by neithernut** [comment]
I finally got something building here. I'll clean up this mess now.

**Comment by neithernut** [comment]
I think this is now ready for the first round of inspection.

File: libimagstore/src/store.rs [hunk 2]
```diff
@@ -147,7 +147,17 @@ impl Store {
 
     /// Delete an entry
     pub fn delete(&self, id: StoreId) -> Result<()> {
-        unimplemented!();
+        let mut entries_lock = self.entries.write();
+        let mut entries = entries_lock.unwrap();
+
+        // if the entry is currently modified by the user, we cannot drop it
+        if entries.get(&id).map( |e| e.is_borrowed() ).unwrap_or(false) {
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: You're hiding an error here. If the `get()` call fails due to the idea not beeing in the store, the error is still `IdLocked`.

Also: Closures are without trailing whitespace: `map(|e| e.is_borrowed())`.

File: libimagstore/src/store.rs [hunk 3]
```diff
@@ -147,7 +147,17 @@ impl Store {
...
+            return Err(StoreError::new(StoreErrorKind::IdLocked, None));
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: You don't need the `return` if you put the things below in the `else{}` branch, which is IMHO a good idea for readability in this case.

File: libimagstore/src/store.rs [hunk 4]
```diff
@@ -147,7 +147,17 @@ impl Store {
...
+        }
+
+        // remove the entry first, then the file
+        entries.remove(&id);
+        remove_file(&id).map_err( |e| { StoreError::new(StoreErrorKind::FileError, Some(Box::new(e))) } )
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Also here, spaces around the closure should be removed.

**Comment by matthiasbeyer** [comment]
Mainly style issues.

AFAIK, the store is `write()`-locked, so if the entry is not borrowed, there is no possibility that it gets borrorwed before we `::remove()` it, right?

File: libimagstore/src/store.rs (hunk 2 above)
Comment (neithernut) [inline libimagstore/src/store.rs]: If the item is not in the store and get returns `None`, we jump over the block returning the error.

**Comment by neithernut** [comment]
That's the idea of keeping around `entries_lock`, yes.

File: libimagstore/src/store.rs (hunk 3 above)
Comment (neithernut) [inline libimagstore/src/store.rs]: Well, conceptually, it's an early return.

File: libimagstore/src/store.rs (hunk 2 above)
Comment (neithernut) [inline libimagstore/src/store.rs]: I'm not sure if I get what you mean.

Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Meh... sorry, you're right.

**Comment by matthiasbeyer** [comment]
Looks good for me now.

For readers of this PR: We always argue with eachother about the early-return thing. So I don't mind merging this without this "fixed" :smile_cat: !

**Comment by matthiasbeyer** [comment]
@homu r+

**Pull request merged**